### PR TITLE
[DRAFT] checkprefix() optimization

### PR DESCRIPTION
**Pull request by stokito**
The PR is not completed and made for discussion. https://github.com/curl/curl/pull/5699#issuecomment-664652440

### CODE REVIEW

**Comment by bagder** [comment]
Do you intend to finish this PR?

**Comment by stokito** [comment]
yes, sorry I'm too busy this days but hope on weekend will finish

File: src/tool_cb_hdr.c [hunk 1]
```diff
@@ -156,7 +156,7 @@ size_t tool_header_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
 
   curl_easy_getinfo(per->curl, CURLINFO_PROTOCOL, &protocol);
   if(hdrcbdata->honor_cd_filename &&
-     (cb > 20) && checkprefix("Content-disposition:", str) &&
+     (cb > 20) && checkprefix("Content-Disposition:", str) &&
```
[the remark this answers is not in the document]
Comment (stokito) [reply src/tool_cb_hdr.c:159]: not really, if char is the same then compassion will continue without additional lower casing. This makes it faster a bit.

File: lib/cookie.c [hunk 2]
```diff
@@ -1135,7 +1136,7 @@ struct CookieInfo *Curl_cookie_init(struct Curl_easy *data,
   c->newsession = newsession; /* new session? */
 
   if(fp) {
-    char *lineptr;
+    const char *lineptr;
```
[the remark this answers is not in the document]
Comment (stokito) [reply lib/cookie.c:1139]: it later used as argument to prefixed_val() but it requires a const pointer. I.e. without this compilation is broken

File: lib/http.c [hunk 3]
```diff
@@ -898,7 +898,7 @@ CURLcode Curl_http_input_auth(struct connectdata *conn, bool proxy,
 
   while(*auth) {
 #ifdef USE_SPNEGO
-    if(checkprefix("Negotiate", auth)) {
+    if(prefixed("Negotiate", auth)) {
```
[the remark this answers is not in the document]
Comment (stokito) [reply lib/http.c:901]: checkprefix is exported and then used from curl tool. To keep backward compatibility and not export the function I created a separate macros. Also it is slightly shorter which is important for long lines in the code.

File: lib/http.c [hunk 4]
```diff
@@ -3746,9 +3748,9 @@ CURLcode Curl_http_readwrite_headers(struct Curl_easy *data,
 
     /* Check for Content-Length: header lines to get size */
     if(!k->http_bodyless &&
-       !data->set.ignorecl && checkprefix("Content-Length:", headp)) {
+       !data->set.ignorecl && prefixed_val("Content-Length:", headp, hv)) {
```
[the remark this answers is not in the document]
Comment (stokito) [reply lib/http.c:3751]: yes, it looks confusing but it still quite simple: once you learned it you understand how to use. I had an idea to make it clearer that `hv` is assigned: pass it by reference:

    prefixed_val("Content-Length:", headp, &hv)

But I decided not to overcomplicate.

File: lib/http.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (bagder) [reply lib/http.c:901]: Is there any reason to keep using the old macro anywhere within libcurl? If not, then I suggest you update the macro anyway and we make the tool code use its own, separate, version of the macro. Changing the name not only means updating 57 lines of code, it also makes readers familiar with the code have to adapt to a new function name unnecessarily.

File: lib/http.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (bagder) [reply lib/http.c:3751]: I insist. Code must not be surprising or contain hidden side effects that require close scrutiny to discover and I think this macro does exactly that.

[the remark this answers is not in the document]
Comment (stokito) [reply lib/http.c:3751]: in fact this is workaround for prohibited assignments inside `if (condition)`. With this macro we at least make this clear

[the remark this answers is not in the document]
Comment (danielgustafsson) [reply lib/http.c:3751]: I agree with @bagder, this is not good practice and not readable. -1 from me.

**Comment by bagder** [comment]
Closing due to inactivity and questionable value

**Pull request closed without merging**